### ISSUE TITLE
@extend and @template don't work well with each other 

### ISSUE THREAD

**Issue by mohsen1** [Microsoft/TypeScript#25101]
When using the `@extend` directive with a generic, TypeScript throws an error.

**TypeScript Version:**  2.9

**Code**
Reproduction repo: https://github.com/mohsen1/extend-ts-bug
```ts
// file: index.js
/**
 * @template {T}
 * @extends {Set<T>}
 */
class MySet extends Set {}

// MySet<number>
const mstNum = new MySet([1, 2, 3]);

// MySet<string>
const mstStr = new MySet(["str1", "str2"]);

```

**Expected behavior:**
No Error and inferred types 

**Actual behavior:**
Error with `@extend`:

```
src/index.js:5:21 - error TS2510: Base constructors must all have the same return type.

5 class MySet extends Set {}
```

Removing extend widens the generic type to `any`

**Related Issues:** 
This issue popped up when working on https://github.com/webpack/webpack/pull/7514

**Comment by sandersn** [comment]
This works except for the bogus error. I think the error just needs to be improved to understand `extends` in JS.

**Comment by sandersn** [comment]
Fix is up at #25111. Turns out lots of places didn't check the `extends` tag and only used the extends heritage clause.
